### PR TITLE
ECC: gen_ecc_key HAL

### PR DESCRIPTION
**Pull request by pascal-brand-st-dev**
Implementation and test of crypto_ops.acipher.gen_ecc_key

Signed-off-by: Pascal Brand <pascal.brand@st.com>

### CODE REVIEW

**Comment by pascal-brand-st-dev** [comment]
xtest 10003 has been added to run the static TA in optee_test; branch ecc_next

**Comment by jenswi-linaro** [comment]
Why don't you put it in the 40xx sequence, it's related to the other asymmetric tests after all.

File: core/sta/ecc_self_tests.c [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+{
+	if (crypto_ops.bignum.compare(p, q) == 0) {
+		EMSG("Keys are equal");
+		return TEE_ERROR_GENERIC;
+	} else {
+		return TEE_SUCCESS;
+	}
+}
+
+static TEE_Result ecc_generate_key_curve(struct ecc_keypair *key1,
```
Comment (jenswi-linaro) [inline core/sta/ecc_self_tests.c]: Can't this be tested through an GP api?

**Comment by pascal-brand-st-dev** [comment]
The aim is to verify the crypto_ops HAL layer for ECC, without waiting for the GP API v1.1 implemented. This is why it is tested this way.

Once GP v1.1 implemented, xtest 40xx will be upgraded with ECC. We'll then check whether or not this static TA is still useful or not.

**Comment by jenswi-linaro** [comment]
OK, I would recommend removing it later since we don't want to start a new trend.
```
 Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>
```

File: core/lib/libtomcrypt/src/tee_ltc_provider.c [hunk 2]
```diff
@@ -1271,6 +1281,93 @@ static TEE_Result alloc_ecc_keypair_dummy(struct ecc_keypair *s,
...
+	case TEE_ECC_CURVE_NIST_P384:
+		key_size_bits = 384;
+		key_size_bytes = 48;
+		break;
+	case TEE_ECC_CURVE_NIST_P521:
+		key_size_bits = 521;
+		key_size_bytes = 65;
+		break;
+	default:
+		return TEE_ERROR_NOT_SUPPORTED;
```
Comment (jbech-linaro) [inline core/lib/libtomcrypt/src/tee_ltc_provider.c]: Would it make sense to set key_size_bits/bytes to 0 just as a precaution?

File: core/sta/ecc_self_tests.c [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+					 uint32_t curve)
+{
+	TEE_Result res;
+	uint32_t key_size_bits;
+
+	/* Generates 2 keys */
+	key1->curve = curve;
+	res = crypto_ops.acipher.gen_ecc_key(key1);
+	if (res != TEE_SUCCESS) {
+		if (curve == TEE_ECC_CURVE_NIST_P521 + 1) {
```
Comment (jbech-linaro) [inline core/sta/ecc_self_tests.c]: I would change this to `if (curve == 0 || curve > TEE_ECC_CURVE_NIST_P521)` instead.

**Comment by jbech-linaro** [comment]
Just two tiny comments, other than that:
`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

PS. Travis errors seems to come from LTC, so I would just ignore them.

**Pull request merged**